Add stat collection to IPTB

davinci26 opened the pull request:
Hey y'all,

Thanks for the project it helped me a lot!

As discussed in IPFS issue board [IPFS Performance #5226](https://github.com/ipfs/go-ipfs/issues/5226), I made some changes to the IPTB framework to generate measure the performance of IPFS and generate performance graphs. In detail I added the following functions in the framework:
- iptb make-topology: This creates a connection graph between the nodes (e.g. star topology, barbell topology). In the topology files empty lines and lines starting with # are disregarded. For non empty line the syntax is `origin:connection 1, connection 2 ...` where origin and connections are specified with their node ID.
- iptb dist -hash: The simulation here distributes a single file from node 0 to every other node in the network. Then it calculates the average time required to download the file, the standard deviation of the time, the maximum time, the minimum time, the duplicate blocks. The results are saved in a generated file called results.json. 

I also added a Python3 script to plot the results that adds an additional optional dependency in the project, Matplotlib. 

Finally I created a readme file, simulation.md, that explains the logic of the simulation. I also added there the response of @whyrusleeping  to the issue [Simulate bad network #50](https://github.com/ipfs/iptb/issues/50) so people would know that there is support for bad network.

I would appreciate your feedback and any improvements suggestions :)

whyrusleeping commented:
Hey @davinci26 Thanks a bunch for this! I hope you don't mind us pulling this apart a bit. We're working on refactoring iptb to be less ipfs centric (giving it the ability to really test *any* libp2p based protocol). And to complete that work, it may require some cutting and pasting :)

I'll leave it to @travisperson and @frrist to take it from here and see how integrating this work could happen.

davinci26 commented:
No problem, feel free to make any changes you want!

I spent quite some time playing with this repo so if you want any help in the refactoring or in the integration, give me a shout :)

davinci26 commented:
Thanks for the review! I totally agree with your suggestions.

From my experience, the problem I had with the `For each` cli command was that it was synchronous. This prohibited me from simulating multiple nodes trying to do the same thing.

Maybe this could also be added as an optional flag in the run command:
`iptb run --async --stats results.json -- ipfs cat <hash> `

I can add the changes as you described them after you merge PR #61 to master, or you can add them directly to PR #61 :).

davinci26 commented:
Hey y'all,
I saw that PR #61 was merged into master.

Can I start working on the `topology` feature and the `statistics metrics` based on the suggestions and comments of @travisperson?

davinci26 commented:
Hey again,

I made the majority of the changes so feel free to review the code. I also added unit tests on the Stats just to be sure there is no stupid miscalculation.

The stats produced are the following:

```
type Stats struct {
	Mean      float64
	Std       float64
	Max       float64
	Quantile3 float64
	Median    float64
	Quantile1 float64
	Min       float64
}

```
Stats Implementation details:

1. In the function `mapWithOutput` I calculate the time elapsed for the command and use it to add it in the `Results struct`
2. I added in the `Build Report` a flag to calculate the stats if the user requests it. Currently, the stats are just printed on the std out. If you are happy with the design I can modify it to also be printed in a file if specified by the user. I print the Stats in `JSON` format for easier parsing by other programs.
3. Also in the `Build Report` I pass an additional string which is the command executed. For me it feels more correct in the case that the output will be automatically parsed by another program to produce statistics (calculate bottlenecks on pipelines). 
4. Right now the majority of commands produce the stats. This could easily be disabled if you dont think its useful.
5.  I assume it would be useful to add stats besides time on the various plugins. However, I am not 100% sure what kind of metrics the plugins you plan to add produce. If you have any reading material on that then I can take a look into it.

Topology Implementation details:

1. Added one function that parses a topology files with the same format as above and then transforms it into the `range format` used internally.

Looking forward to your comments 😃

dgrisham commented on commands/utils.go:
```diff
@@ -225,91 +226,73 @@ func buildReport(results []Result, command string, statsFlag bool) error {
-		c1 = (length-1)/2 + 1
-		c2 = c1 - 1
+// Print individual results to JSON format
+func printIndivudualResults(nodeTime []float64, nodeID []int, metric string) {
```
Think that's a typo and should be `printIndividualResults` ;)

dgrisham commented on commands/utils.go:
```diff
@@ -195,13 +203,96 @@ func buildReport(results []Result) error {
+the strings are not castable to float
+the function returns an error
+*/
+func substractArrays(lhs []string, rhs []string) ([]float64, error) {
```
One more typo, `subtractArrays`*

davinci26 replied:
I will fix all of them :) Thnx a lot for the review, I totally agree details matter a lot !

dgrisham replied:
No worries! And one of the comments I made got marked as outdated for some reason, but `printIndividualResults` is typo'd as `printIndivudualResults`

dgrisham commented on commands/stats.go:
```diff
@@ -0,0 +1,101 @@
+
+	// Calculate the min, max mean, std with a single pass of the array
+	sum := 0.
+	sumSquarred := 0.
```
`sumSquared` (just one r)

dgrisham commented on commands/stats.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+//BuildStats generates a stats struct from an input float64 slice
+//It returns an error if the slice is empty
```
These two lines are missing a space after the `//` to be consistent with the others (sorry, details stand out to me haha)

dgrisham commented on README.md:
````diff
@@ -26,7 +26,7 @@ $ iptb shell 4
 hey!
 ```
-
+Available plugins now: Local IPFS node (plugin_name: localipfs), Docker IPFS node (plugin_name: dockeripfs)
````
Would also be nice to be able to list available plugins via the CLI. Doesn't have to be in this PR though, just mentioning.

dgrisham commented on commands/connect.go:
```diff
@@ -23,13 +27,17 @@ The connect command allows for connecting sets of nodes together.
-There are three variants of the command. It can accept no arugments,
-a single argument, or two arguments. The no argument and single argument
-expands out to the two argument usage.
+There are four variants of the command. It can accept no arugments,
```
arguments*

dgrisham commented on commands/connect.go:
```diff
@@ -23,13 +27,17 @@ The connect command allows for connecting sets of nodes together.
-expands out to the two argument usage.
+There are four variants of the command. It can accept no arugments,
+a single argument, two arguments or a file. The no argument and single argument
+expands out to the two argument usage. The file argument, parses the file and exapnds it
```
expands* (the second one)

davinci26 commented on main.go:
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 func loadPlugins(dir string) error {
+
```
The code formatting tool decided to add an empty line there. I have manually checked the files and they are the same.

davinci26 commented on commands/run.go:
```diff
@@ -54,11 +65,51 @@ var RunCmd = cli.Command{
+		nodeMetricValue[i] = value
+	}
+	return nodeMetricValue, nil
+}
```
Here I opted to collect the metric before and after the command is executed in `run.go`. I think with this approach the code is cleaner and makes more sense.
I could add it in the `mapWithOutput` function but since this function is using locks the more operations I do inside the locks the more the overhead.

I would love to hear your thoughts on which design you think is more useful for the metrics you plan to implement.

davinci26 commented on commands/utils.go:
```diff
@@ -195,13 +203,97 @@ func buildReport(results []Result) error {
+the function returns an error
+TODO: This could be more generic
+*/
+func subtractArrays(lhs []string, rhs []string) ([]float64, error) {
```
As it is implemented now the metrics are collected as follows:
Let metric <m1> be collected, then we collect m1<sub>before</sub> before the execution of the run command and m1<sub>after</sub> after. Then we calculate the statistics based on Δm1 = (m1<sub>after</sub> - m1<sub>before</sub>).

davinci26 commented:
The last commit reflects the discussion in #77 #76 . The user can run 
`ipfs run 0 --encoding JSON <args>` or `ipfs run 0 --encoding text <args>` and get the outpout in the said encoding.
The JSON encoding transforms the output (including the plugin stderr and stdout) into a JSON. This way the user can specify he/she prefers a human-readable or a machine-readable output.

davinci26 commented:
I handled the error cases mentioned and also the formatting is as specified when I append to a file.

davinci26 commented:
Hey @frrist,

I was thinking about this as well, that this PR is getting too big and for no reason, as there was no design discussion prior to the beginning of the work.  It was also an omission from my side for not opening an issue prior to the PR. Feel free to decline the PR.

I have written my ideas in #77 under general thoughts and I am more than happy to help in designing meaningful event logs.

whyrusleeping commented:
I'm really excited that iptb is getting so much attention. I love a lot of the direction here, let's get things designed nicely and march onwards :) Thanks @davinci26 for pushing!

One thing I always try to keep in mind in the design of tools is the unix philosophy, primarily "do one thing well". What that means here specifically, I'm not sure, but I'm thinking that we may end up adding a companion tool to iptb for metrics collection, and just ensuring that iptb doesnt prevent us from getting the information we need. This would allow us to more easily tune the stats we collect for each usecase. (similar design philosophy went into the plugins architecture)
